### ISSUE TITLE
[Bug]: reshape op output is incorrect in dynamic shape

### ISSUE THREAD

**Issue by ecneil** [openvinotoolkit/openvino#21051]
### OpenVINO Version

2023.1

### Operating System

Ubuntu 22.04 (LTS)

### Device used for inference

CPU and GPU

### Framework

None

### Model used

None

### Issue description

We encountered an output problem with one of our models.
After isolating the problem, we found that the reshape operation returned incorrect output when a chain of inferrence of a certain size were applied.

The output results are different, but similar problem occur on CPU and GPU respectively.

### Step-by-step reproduction

Execute the following code.

```
import openvino.runtime as ovr
import numpy as np

def create_reshape_model():
    input_ = ovr.opset11.parameter([1, 3, -1, -1], ovr.Type.u8)
    input_shape = ovr.opset11.shape_of(input_)
    input_shape_2 = ovr.opset11.gather(input_shape, 2, 0, 0)
    input_shape_3 = ovr.opset11.gather(input_shape, 3, 0, 0)
    flat_shape = ovr.opset11.multiply(input_shape_2, input_shape_3)
    reshape1_new_shape = ovr.opset11.concat([np.array([1, 3]), ovr.opset11.unsqueeze(flat_shape, [0])], 0)
    reshape1 = ovr.opset11.reshape(input_, reshape1_new_shape, True, name="reshape1")
    reshape2 = ovr.opset11.reshape(reshape1, input_shape, True, name="reshape2")
    output = ovr.opset11.result(reshape2)
    model = ovr.Model([output], [input_], "model")
    model.outputs[0].tensor.names = {"output"}
    return model

core = ovr.Core()
model = create_reshape_model()
input_name = model.inputs[0].any_name
output_name = model.outputs[0].any_name
compiled_model = core.compile_model(model, "AUTO")
infer_req = compiled_model.create_infer_request()

input_shape = (
    (320, 320),
    (160, 320),
    (640, 80)
)

for h, w in input_shape * 2:
    input_data = np.random.randint(0, 255, (1, 3, h, w), np.uint8)
    pred = infer_req.infer({input_name: input_data})[output_name]
    print(f"output shape is {'correct' if pred.shape == input_data.shape else 'incorrect'}. {input_data.shape=}, {pred.shape=}")
```

### Relevant log output

on cpu
```
output shape is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape is correct. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 640, 80)
output shape is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape is incorrect. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 160, 320)
```

on gpu
```
output shape is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape is incorrect. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 160, 320)
output shape is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape is incorrect. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 160, 320)
```

### Issue submission checklist

- [X] I'm reporting an issue. It's not a question.
- [X] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [X] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by mlukasze** [comment]
thanks for information.
let us check it out, we will back to you soon.

**Comment by praasz** [comment]
I can confirm that issue exist on CPU, GPU plugin but there is no issue on `TEMPLATE` plugin which calculates shape as expected.

It looks like issue in GPU and CPU is similar when consecutive inference occurs, the plugins at some point don't recognize that shapes changed and it's required to re-calculate them. Results that output shape from previous inference has been used. 

It looks like is related with `reshape1_new_shape` where last dimension is dynamic and it value is product of `input_shape` dimensions (h * w), which gives same results for [160, 320] and [640, 80] dimensions.

 - In CPU plugin removing this [check](https://github.com/openvinotoolkit/openvino/blob/b380eb06ee10e67dcc0f3723b8cfbb9a26e80ed8/src/plugins/intel_cpu/src/nodes/reshape.cpp#L68) in reshape node will solve this issue. Could CPU team check if this is correct solution?
-  In GPU plugin in reshape node at [line](https://github.com/openvinotoolkit/openvino/blob/b380eb06ee10e67dcc0f3723b8cfbb9a26e80ed8/src/plugins/intel_gpu/src/graph/reshape.cpp#L135) the output shape is calculated correctly but not used as inference result shape. Could GPU team propose solution how to update output shape correctly?

**Comment by maxnick** [comment]
@praasz , thank you for the investigation!!! Regarding your proposal for the CPU plugin, at first glance removing the check may have a negative performance impact, so the solution should be more sophisticated. @tiger100256-hu, could you please take a look at the issue?

**Comment by ecneil** [comment]
Do you plan to include a fix for the CPU plugin in 2023.3?
Also, has any progress been made so far on the GPU plugin fixes?

**Comment by wenjiew** [comment]
@ecneil Yes on the CPU part, this will be included in 2023.3 release. 
@p-durandin to comment on the GPU side.

**Comment by avitial** [comment]
Tested this use case on 2024.4 version and the issue seems to have been addressed, both CPU/GPU output the same results. Closing this as it has been addressed. Feel free to reopen and ask additional questions related to this topic if the issue persists.

```
$ python script.py
output shape for CPU is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape for CPU is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape for CPU is correct. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 640, 80)
output shape for CPU is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape for CPU is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape for CPU is correct. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 640, 80)
```
```
$ python script.py
output shape for GPU is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape for GPU is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape for GPU is correct. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 640, 80)
output shape for GPU is correct. input_data.shape=(1, 3, 320, 320), pred.shape=(1, 3, 320, 320)
output shape for GPU is correct. input_data.shape=(1, 3, 160, 320), pred.shape=(1, 3, 160, 320)
output shape for GPU is correct. input_data.shape=(1, 3, 640, 80), pred.shape=(1, 3, 640, 80)
```
